diff/walking/differ.go

```diff
@@ -53,9 +53,9 @@ type walkingDiff struct {
 
 var emptyDesc = ocispec.Descriptor{}
 
-// NewWalkingDiff is a generic implementation of diff.Differ.
+// NewWalkingDiff is a generic implementation of diff.DiffApplier.
 // NewWalkingDiff is expected to work with any filesystem.
-func NewWalkingDiff(store content.Store) (diff.Differ, error) {
+func NewWalkingDiff(store content.Store) (diff.DiffApplier, error) {
```

Comment:
What does the `walkingDiff` have to do with the `Apply()` method it implements. Eg. what is being walked in that method.

Reply:
The apply algorithm and diff algorithm are really unrelated. The Apply method here is just applying on a mounted file system. Same with diff, it is just mounting and computing the diff. I am not convinced this PR is adequately cleaning up nomenclature confusion around diff and apply. The term diff is being used as a term for managing diffs, including generation and application of diff, splitting the 2 in this way is kind of making it more confusing...

Reply:
> The term diff is being used as a term for managing diffs, including generation and application of diff

For the split, there are different parts of the system that only either need an `apply` or `diff` method so it is bad that to use these subsystems, one has to implement both.

For my question, the "WalkingDiff" name should mean that the diff is made by recursively walking both sides and comparing items one by one(in Docker this is called `NaiveDiff`). If the algorithm would be different, for example "NewNativeDiff()/NewBtrfsDiff()` for overlay/aufs/btrfs, the same interface would be implemented by another algorithm. But none of that separation seems to apply to the `Apply()` method that would be exactly the same for these different variants.

Reply:
@tonistiigi Should we split the `diff` gRPC service as well?
If so, we need to do that before v1.0 GA.

Reply:
@AkihiroSuda IANAM, but I don't think that is super critical. It's fine to combine things as "diffing-related-utilities". But there is no such thing as `WalkingApplier` and implementors should not need to implement all the utilities just to use one.

Reply:
@tonistiigi @dmcgowan 

Can this PR be merged as-is?
In follow-up PR, we may split `diff/walking` implementation to `diff/appliers/basic`.